query-algebrizer/src/validate.rs

```diff
@@ -0,0 +1,234 @@
+                                        [?artist :artist/gender :artist.gender/female]))]"#;
+        let parsed = parse_find_string(query).unwrap();
+        let clauses = valid_or_join(parsed, UnifyVars::Implicit);
+
+        // Let's do some detailed parse checks.
+        let mut arms = clauses.into_iter();
+        match (arms.next(), arms.next(), arms.next()) {
+            (Some(left),
+            Some(right),
+            None) => {
+                assert_eq!(
+                    left,
+                    OrWhereClause::Clause(WhereClause::Pattern(Pattern {
+                        source: None,
+                        entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?artist"))),
+                        attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("artist", "type")),
+                        value: PatternValuePlace::IdentOrKeyword(NamespacedKeyword::new("artist.type", "group")),
+                        tx: PatternNonValuePlace::Placeholder,
+                    })));
+                assert_eq!(
+                    right,
+                    OrWhereClause::And(
+                        vec![
+                            WhereClause::Pattern(Pattern {
+                                source: None,
+                                entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?artist"))),
+                                attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("artist", "type")),
+                                value: PatternValuePlace::IdentOrKeyword(NamespacedKeyword::new("artist.type", "person")),
+                                tx: PatternNonValuePlace::Placeholder,
+                            }),
+                            WhereClause::Pattern(Pattern {
+                                source: None,
+                                entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?artist"))),
+                                attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("artist", "gender")),
+                                value: PatternValuePlace::IdentOrKeyword(NamespacedKeyword::new("artist.gender", "female")),
+                                tx: PatternNonValuePlace::Placeholder,
+                            }),
+                        ]));
+            },
+            _ => panic!(),
+        };
+    }
+
+    /// Test that an `or` with differing variable sets in each arm will fail to validate.
+    #[test]
+    fn test_invalid_implicit_or() {
+        let query = r#"[:find [?artist ...]
+                        :where (or [?artist :artist/type :artist.type/group]
+                                   [?artist :artist/type ?type])]"#;
+        let parsed = parse_find_string(query).unwrap();
+        match parsed.where_clauses.into_iter().next().unwrap() {
+            WhereClause::OrJoin(or_join) => assert!(validate_or_join(&or_join).is_err()),
```

Comment:
This style can be quite difficult to debug when you break the tests, since you'll get unhelpful panics from the `unwrap` invocations.